File: src/commands/profile/list.ts
import chalk from "chalk"
import fs from "fs"
import path from "path"
import getCache from "src/utils/cache"
import getConfig from "src/utils/config"
import jarVersion from "src/utils/jar"

export type Args = {}

export default async function profileList(args: Args) {
	const config = getConfig(),
		cache = getCache(),
		profiles = new Set((await fs.promises.readdir('.mcvcli.profiles').catch(() => []))
			.filter((profile) => fs.existsSync(path.join('.mcvcli.profiles', profile, '.mcvcli.json'))))

	profiles.add(config.data.profileName)

	const profileVersions: Record<string, string> = {},
		start = performance.now()

	console.log('checking versions...')

	const data = await Promise.all([...profiles].map(async(profile) => {
		const profileConfig = getConfig(profile === config.data.profileName ? undefined : profile),
			version = await jarVersion(profile === config.data.profileName ? profileConfig.data.jarFile : path.join('.mcvcli.profiles', profile, profileConfig.data.jarFile), cache)

		return version.type === 'unknown' ? 'unknown' : `${version.type} ${version.minecraftVersion}`
	}))

	for (let i = 0; i < profiles.size; i++) profileVersions[[...profiles][i]] = data[i]

	console.log('checking versions... done', chalk.gray(`(${(performance.now() - start).toFixed(2)}ms)`), '\n')

	console.log('profiles:')
	if (!profiles.size) console.log('  (none)')

	for (const profile of profiles) {
		console.log('  ', profile === config.data.profileName ? chalk.cyan(profile) : profile, chalk.gray(`(${profileVersions[profile]})`))
	}

	console.log('current profile:', chalk.cyan(config.data.profileName))
}

File: src/commands/profile/list.rs
use crate::{config, jar, profiles};

use clap::ArgMatches;
use colored::Colorize;
use std::path::Path;

pub async fn list(matches: &ArgMatches) -> Result<i32, anyhow::Error> {
    let include_version = matches
        .get_one::<bool>("include_version")
        .expect("has default value");
    let config = config::Config::new(".mcvcli.json", false);

    println!("{}", "listing profiles...".bright_black());

    let mut list = profiles::list();
    list.push(config.profile_name.clone());

    let mut futures = Vec::new();

    if *include_version {
        for profile in list.iter() {
            let directory = if *profile != config.profile_name {
                format!(".mcvcli.profiles/{profile}")
            } else {
                String::from(".")
            };

            futures.push(async move {
                let profile_config = config::Config::new(
                    &Path::new(&directory).join(".mcvcli.json").to_string_lossy(),
                    false,
                );

                jar::detect(&directory, &profile_config).await
            });
        }
    }

    let results = futures::future::join_all(futures).await;

    println!(
        "{} {}",
        "listing profiles...".bright_black(),
        "DONE".green().bold()
    );

    for (i, profile) in list.into_iter().enumerate() {
        println!();

        let directory = if *profile != config.profile_name {
            format!(".mcvcli.profiles/{profile}")
        } else {
            String::from(".")
        };

        let profile_config = config::Config::new(
            &Path::new(&directory).join(".mcvcli.json").to_string_lossy(),
            false,
        );

        println!(
            "{} {}",
            profile.cyan().bold().underline(),
            if *profile == config.profile_name {
                "(current)".green()
            } else {
                String::new().green()
            }
        );

        println!(
            "  {} {}",
            "jar file:    ".bright_black(),
            profile_config.jar_file.cyan()
        );
        println!(
            "  {} {}",
            "java version:".bright_black(),
            profile_config.java_version.to_string().cyan()
        );
        println!(
            "  {} {}",
            "ram (mb):    ".bright_black(),
            profile_config.ram_mb.to_string().cyan()
        );

        if *include_version {
            let detected = results
                .get(i)
                .ok_or_else(|| anyhow::anyhow!("missing profile detection result"))?
                .as_ref();

            if let Some(([build, latest], versions, modpack)) = detected {
                println!("  {}", "version:".bright_black());
                println!("    {} {}", "type:   ".bright_black(), build.r#type.cyan());
                println!(
                    "    {} {} {}",
                    "version:".bright_black(),
                    build
                        .version_id
                        .as_ref()
                        .unwrap_or(
                            build
                                .project_version_id
                                .as_ref()
                                .unwrap_or(&"unknown".to_string())
                        )
                        .cyan(),
                    if jar::is_latest_version(build, versions) {
                        "(latest)".green()
                    } else {
                        "(outdated)".red()
                    }
                );
                println!(
                    "    {} {} {}",
                    "build:  ".bright_black(),
                    build.name.cyan(),
                    if build.uuid == latest.uuid {
                        "(latest)".green()
                    } else {
                        "(outdated)".red()
                    }
                );

                if let Some(modpack) = modpack {
                    println!("  {}", "installed modpack:".bright_black());
                    println!(
                        "    {} {}",
                        "name:       ".bright_black(),
                        modpack.title.cyan()
                    );
                    println!(
                        "    {} {}",
                        "description:".bright_black(),
                        modpack.description.cyan()
                    );
                    println!(
                        "    {} {}",
                        "project id: ".bright_black(),
                        modpack.id.as_deref().unwrap_or("unknown").cyan()
                    );
                    if let Some(version) = config.modpack_version.as_ref() {
                        println!(
                            "    {} {} {}",
                            "version id: ".bright_black(),
                            version.cyan(),
                            if modpack.versions.last() == Some(version) {
                                "(latest)".green()
                            } else {
                                "(outdated)".red()
                            }
                        );
                    }
                    println!(
                        "    {} {}",
                        "downloads:  ".bright_black(),
                        modpack.downloads.to_string().cyan()
                    );
                    println!();
                }
            } else {
                println!("  {} {}", "version:     ".bright_black(), "unknown".cyan());
            }
        }
    }

    Ok(0)
}
